src/sentry/static/sentry/app/views/performance/compare/utils.tsx

```diff
@@ -0,0 +1,771 @@
+}: {
+  parent_span_id: SpanId;
+  baseChildren: Array<SpanType>;
+  regressionChildren: Array<SpanType>;
+}): {
+  comparablePairs: Array<ComparableSpan>;
+  children: Array<DiffSpanType>;
+} {
+  // invariant: the parents of baseChildren and regressionChildren are matched spans
+
+  // for each child in baseChildren, pair them with the closest matching child in regressionChildren
+
+  const comparablePairs: Array<ComparableSpan> = [];
+  const children: Array<DiffSpanType> = [];
+
+  const remainingRegressionChildren = [...regressionChildren];
+
+  for (const baselineSpan of baseChildren) {
+    // reduce remainingRegressionChildren down to spans that are applicable candidate
+    // of spans that can be paired with baselineSpan
+
+    const candidates = remainingRegressionChildren.reduce(
+      (
+        acc: Array<{regressionSpan: SpanType; index: number}>,
+        regressionSpan: SpanType,
+        index: number
+      ) => {
+        if (matchableSpans({baselineSpan, regressionSpan})) {
+          acc.push({
+            regressionSpan,
+            index,
+          });
+        }
+
+        return acc;
+      },
+      []
+    );
+
+    if (candidates.length === 0) {
+      children.push({
+        comparisonResult: 'baseline',
+        baselineSpan,
+      });
+      continue;
+    }
+
+    // the best candidate span is one that has the closest start timestamp to baselineSpan;
+    // and one that has a duration that's close to baselineSpan
+
+    const baselineSpanDuration = Math.abs(
+      baselineSpan.timestamp - baselineSpan.start_timestamp
+    );
+
+    candidates.sort(({regressionSpan: thisSpan}, {regressionSpan: otherSpan}) => {
+      // calculate the deltas of the start timestamps relative to baselineSpan's
+      // start timestamp
+
+      const deltaStartTimestampThisSpan = Math.abs(
+        thisSpan.start_timestamp - baselineSpan.start_timestamp
+      );
+
+      const deltaStartTimestampOtherSpan = Math.abs(
+        otherSpan.start_timestamp - baselineSpan.start_timestamp
+      );
+
+      // calculate the deltas of the durations relative to the baselineSpan's
+      // duration
+
+      const thisSpanDuration = Math.abs(thisSpan.timestamp - thisSpan.start_timestamp);
+      const otherSpanDuration = Math.abs(otherSpan.timestamp - otherSpan.start_timestamp);
+
+      const deltaDurationThisSpan = Math.abs(thisSpanDuration - baselineSpanDuration);
+      const deltaDurationOtherSpan = Math.abs(otherSpanDuration - baselineSpanDuration);
+
+      const thisSpanScore = deltaDurationThisSpan + deltaStartTimestampThisSpan;
+      const otherSpanScore = deltaDurationOtherSpan + deltaStartTimestampOtherSpan;
+
+      if (thisSpanScore < otherSpanScore) {
+        // sort thisSpan before otherSpan
+        return -1;
+      }
+
+      if (thisSpanScore > otherSpanScore) {
+        // sort thisSpan after otherSpan
+        return 1;
+      }
+
+      return 0;
+    });
+
+    const {regressionSpan, index} = candidates[0];
```

Comment:
Is this just getting the span with the minimal delta from the start? If so, you should be able to simply iterate it and get the min in O(n)

Reply:
Should be addressed in https://github.com/getsentry/sentry/pull/18610/commits/34dc6bc59898a732bbcd553e26582a276987c183